### ISSUE TITLE
TPH Mapping: ApplyConfiguration specific order required?

### ISSUE THREAD

**Issue by AEonWho** [aspnet/EntityFrameworkCore#18385]
I'm trying to migrate from efcore 2.2 to 3.0 and ran into a issue with our Entities that have inheritance defined. It seems that now the order in which our mappings are configured is important, and since it worked before, it feels like a bug.

I simplified our structure as much as possible to reproduce the issue.

### Steps to reproduce
Entity classes:

``` C#
public abstract class BaseData
{
    [Key]
    public int Id { get; set; }
}

public abstract class SubBaseData : BaseData
{
}

public class MainData : BaseData
{
}

public class SubData : SubBaseData
{
}
```

Mappings:
``` C#
public class BaseDataMap : IEntityTypeConfiguration<BaseData>
{
    public void Configure(EntityTypeBuilder<BaseData> builder)
    {
        builder.HasDiscriminator<int>("TypeDiscriminator")
               .HasValue<MainData>(1)
               .HasValue<SubData>(2);
    }
}

public class MainDataMap : IEntityTypeConfiguration<MainData>
{
    public void Configure(EntityTypeBuilder<MainData> builder)
    {
        // We have properties defined, but it also fails when it's empty.
    }
}

public class SubBaseDataMap : IEntityTypeConfiguration<SubBaseData>
{
    public void Configure(EntityTypeBuilder<SubBaseData> builder)
    {
        // We have properties defined, but it also fails when it's empty.
    }
}
```

DBContext (this one fails when trying to query DataEntries)
``` C#
public DbSet<BaseData> DataEntries { get; set; }

protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    base.OnModelCreating(modelBuilder);

    modelBuilder.ApplyConfiguration(new BaseDataMap());
    modelBuilder.ApplyConfiguration(new MainDataMap());
    modelBuilder.ApplyConfiguration(new SubBaseDataMap());
}
```
Exception:
```
Unhandled exception. System.InvalidOperationException: The entity type 'SubData' is part of a hierarchy, but does not have a discriminator value configured.
```

When i switch the order, so that my base mapping with the discriminator definition is last, it works.
``` C#
modelBuilder.ApplyConfiguration(new MainDataMap());
modelBuilder.ApplyConfiguration(new SubBaseDataMap());
modelBuilder.ApplyConfiguration(new BaseDataMap());
```

For me it seems that the `SubBaseDataMap` mapping is somehow overriding the discriminator/inheritance defined in the `BaseDataMap`.

Here is my demo wich demonstrates this with the InMemory Provider, i also tested it on an local sqlserver instance.
[Program.zip](https://github.com/aspnet/EntityFrameworkCore/files/3731194/Program.zip)

### Further technical details

EF Core version: 3.0.0 (current NuGet)
Database provider: Microsoft.EntityFrameworkCore.InMemory, Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.0
Operating system: Windows 10
IDE: Visual Studio 2019 16.3.5

**Comment by Rubenisme** [comment]
Thanks for finding out it was the order, we also had this problem and I hadn't tried changing the orders of the mappings, although in another case it worked so it kind of makes sense that it was this.

**Comment by bdaniel7** [comment]
Is this fixed for EF Core 3.1.x? This used to work in EF Core 2.2.
I have a bunch of configuration classes that implement IEntityTypeConfiguration<> and are loaded and applied automatically via reflection. 
Now I have to sort the configuration files so the base class is applied last?

**Comment by ajcvickers** [comment]
@bdaniel7 This was fixed in 3.1.0, as indicated by the milestone.

**Comment by bdaniel7** [comment]
Ok, I removed the IEntityTypeConfiguration<> for the derived types and then it worked.
